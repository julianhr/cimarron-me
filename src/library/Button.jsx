import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const NoOp = () => {}

const Btn = styled.button`
  border: 2px solid ${props => props.theme.colors.primary.dark};
  border-radius: 5px;
  background: transparent;
  padding: 4px 16px;
  color: ${props => props.theme.colors.primary.dark};
  font-size: 1.3em;
  height: 44px;
  min-width: 140px;

  :hover {
    cursor: pointer;
  }

  :disabled {
    border: 2px solid ${props => props.theme.colors.primary.dim};
    color: ${props => props.theme.colors.primary.dim};
  }
`

function Button({ children, type, btnStyle, onClick, fwdRef }) {
  return (
    <Btn
      ref={fwdRef}
      type={type}
      onClick={onClick}
      css={btnStyle}
    >
      {children}
    </Btn>
  )
}

Button.propTypes = {
  btnStyle: PropTypes.object,
  children: PropTypes.string.isRequired,
  fwdRef: PropTypes.object,
  onClick: PropTypes.func,
  type: PropTypes.string,
}

Button.defaultProps = {
  btnStyle: {},
  onClick: NoOp,
  type: 'button',
}

export default Button
