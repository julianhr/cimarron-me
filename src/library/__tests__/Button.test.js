import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { create } from 'react-test-renderer'
import emotionSerializer from 'jest-emotion'

import Button from '../Button'
import MockApp from '~/__tests__/__mocks__/MockApp'


const testProps = {
  btnStyle: { color: 'green' },
  fwdRef: React.createRef(),
  onClick: () => {},
  type: 'submit',
}

const buttonText = 'Hello!'

const createApp = (props, text) => create(
  <MockApp>
    <Button {...props}>
      {text}
    </Button>
  </MockApp>
)

const mountApp = (props, text) => mount(
  <MockApp>
    <Button {...props}>
      {text}
    </Button>
  </MockApp>
)

describe('Button', () => {
  describe('props', () => {
    test('btnStyle not required', () => {
      const props = {...testProps}
      delete props.btnStyle
      expect(() => createApp(props, buttonText)).not.toThrow()
    })

    test('children is required', () => {
      expect(() => create(
        <MockApp><Button {...testProps} /></MockApp>)).toThrow()
    })

    test('fwdRef is not required', () => {
      expect(() => createApp(testProps, buttonText)).not.toThrow()
    })

    test('onClick is not required', () => {
      const props = {...testProps}
      delete props.onClick
      expect(() => createApp(props, buttonText)).not.toThrow()
    })

    test('type is not required', () => {
      const props = {...testProps}
      delete props.type
      expect(() => createApp(props, buttonText)).not.toThrow()
    })
  })

  describe('snapshots', () => {
    test('with all props', () => {
      expect.addSnapshotSerializer(emotionSerializer)
      const wrapper = mountApp(testProps, buttonText)
      expect(wrapper).toMatchSnapshot()
      wrapper.unmount()
    })

    test('with required props', () => {
      const props = { onClick: testProps.onClick }
      const wrapper = mountApp(props, buttonText)

      expect.addSnapshotSerializer(emotionSerializer)
      expect(wrapper).toMatchSnapshot()
      wrapper.unmount()
    })
  })

  it('calls onClick callback when button is clicked', () => {
    const spy = jest.spyOn(testProps, 'onClick')
    const wrapper = mountApp(testProps, buttonText)
    wrapper.find('button').simulate('click')
    expect(spy).toHaveBeenCalledTimes(1)
  })
})
