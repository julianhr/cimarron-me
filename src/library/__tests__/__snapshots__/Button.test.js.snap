// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshots with all props 1`] = `
Array [
  <ForwardRef(render)
    styles={[Function]}
  />,
  .emotion-2 {
  color: green;
}

.emotion-0 {
  border: 2px solid #9B1518;
  border-radius: 5px;
  background: transparent;
  padding: 4px 16px;
  color: #9B1518;
  font-size: 1.2em;
  height: 44px;
  min-width: 140px;
  color: green;
}

.emotion-0:hover {
  cursor: pointer;
}

.emotion-0:disabled {
  border: 2px solid #FFA8AA;
  color: #FFA8AA;
}

<Btn
    className="emotion-2"
    onClick={[Function]}
    type="submit"
  >
    <button
      className="emotion-0 emotion-1"
      onClick={[Function]}
      type="submit"
    >
      Hello!
    </button>
  </Btn>,
]
`;

exports[`Button snapshots with required props 1`] = `
Array [
  <ForwardRef(render)
    styles={[Function]}
  />,
  .emotion-0 {
  border: 2px solid #9B1518;
  border-radius: 5px;
  background: transparent;
  padding: 4px 16px;
  color: #9B1518;
  font-size: 1.2em;
  height: 44px;
  min-width: 140px;
}

.emotion-0:hover {
  cursor: pointer;
}

.emotion-0:disabled {
  border: 2px solid #FFA8AA;
  color: #FFA8AA;
}

<Btn
    className="emotion-2"
    onClick={[Function]}
    type="button"
  >
    <button
      className="emotion-0 emotion-1"
      onClick={[Function]}
      type="button"
    >
      Hello!
    </button>
  </Btn>,
]
`;
